cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,78 +282,370 @@ func copyFromGlobHelper(src, dst, globEx string) error {
 	return nil
 }
 
-func generateMountsModeRecover(mst *initramfsMountsState) error {
-	// steps 1 and 2 are shared with install mode
-	if err := generateMountsCommonInstallRecover(mst); err != nil {
+type recoverDegradedState struct {
+	// TODO: make these values enums?
+	// DataKey is which key was used to unlock ubuntu-data (if any). Valid
+	// values are:
+	// - "run" for the normal run mode key object
+	// - "fallback" for the fallback recover mode specific key object
+	// - "" for either unencrypted case, or if we were unable to unlock it
+	//   (in the case we failed to unlock it, but we know it is there, then
+	//   )
+	DataKey string `json:"data-key"`
+	// TODO: make these values enums?
+	// DataState is the state of the ubuntu-data mountpoint, it can be:
+	// - "mounted" for mounted and available
+	// - "enc-not-found" for state where we found an ubuntu-data-enc
+	//   partition and tried to unlock it, but failed _before mounting it_.
+	//   Errors mounting are identified differently.
+	// TODO:UC20: "other" is a really bad term here replace with something better
+	// - "other" for some other error where we couldn't identify if there's an
+	//    ubuntu-data (or ubuntu-data-enc) partition at all
+	DataState string `json:"data-state"`
+	// SaveKey is which key was used to unlock ubuntu-save (if any). Same values
+	// as DataKey.
+	SaveKey string `json:"save-key"`
+	// SaveState is the state of the ubuntu-save mountpoint, it can be in any of
+	// the states documented for DataState, with the additional state of:
+	// - "not-needed" for state when we don't have ubuntu-save, but we don't need it (unencrypted only)
+	SaveState string `json:"save-state"`
+	// HostLocation is the location where the host's ubuntu-data is mounted
+	// and available. If this is the empty string, then the host's
+	// ubuntu-data is not available anywhere.
+	HostLocation string `json:"host-location"`
+}
```

Comment:
We could add `ErrorLog []string` here to capture each error that happens along the way.